test/JoinTest.cpp

```diff
@@ -0,0 +1,429 @@
+ * join function and compares the results with the given expected content.
+ *
+ * @tparam JOIN_FUNCTION is used to allow the transfer of any type of
+ *  lambda function, that could contain a join function. You never have to
+ *  actually specify this parameter, just let inference do its job.
+ *
+ * @param testSet is an array of normalJoinTests, that describe what tests to
+ *  do. For an explanation, how to read normalJoinTest, see the definition.
+ * @param func the function, that will be used for joining the two tables
+ *  together. Look into src/engine/Join.h, or this file, for how it should look like.
+ */
+template<typename JOIN_FUNCTION>
+void goThroughSetOfTestsWithJoinFunction(
+      const std::vector<normalJoinTest>& testSet,
+      JOIN_FUNCTION func,
+      ad_utility::source_location l = ad_utility::source_location::current()
+    ) {
+  // For generating better messages, when failing a test.
+  auto trace{generateLocationTrace(l, "goThroughSetOfTestsWithJoinFunction")};
+
+  for (normalJoinTest const& test: testSet) {
+    IdTable resultTable{useJoinFunctionOnIdTables(test.leftInput,
+        test.rightInput, func)};
+
+    compareIdTableWithExpectedContent(resultTable, test.expectedResult, test.resultMustBeSortedByJoinColumn, test.leftInput.joinColumn);
+  }
+}
+
+void runTestCasesForAllJoinAlgorithm(std::vector<normalJoinTest> testSet,
+    ad_utility::source_location l = ad_utility::source_location::current()) {
+  // For generating better messages, when failing a test.
+  auto trace{generateLocationTrace(l, "runTestCasesForAllJoinAlgorithm")};
+ 
+  // All normal join algorithm defined as lambda functions for easier handing
+  // over to helper functions.
+  Join J{Join::InvalidOnlyForTestingJoinTag{}};
+  auto HashJoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.hashJoin<A, B, C>(AD_FWD(args)...);
+  };
+  auto JoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
+
+  // For sorting IdTables by their join column.
+  auto sortByJoinColumn = [](IdTable& idTable, size_t joinColumn) {
+    /*
+     * TODO Introduce functionality to the IdTable-class, so that std::ranges::sort
+     * can be used instead of std::sort. Currently it seems like the iterators
+     * , produced by IdTable, aren't the right type.
+     */
+    std::sort(idTable.begin(), idTable.end(), [&joinColumn](const auto& row1,
+          const auto& row2) {return row1[joinColumn] < row2[joinColumn];});
+  };
+
+  // Random shuffle both tables, run hashJoin, check result.
+  for (normalJoinTest& testCase: testSet) {
+    randomShuffle(testCase.leftInput.idTable.begin(), testCase.leftInput.idTable.end());
+    randomShuffle(testCase.rightInput.idTable.begin(), testCase.rightInput.idTable.end());
+    testCase.resultMustBeSortedByJoinColumn = false;
+  }
+  goThroughSetOfTestsWithJoinFunction(testSet, HashJoinLambda);
```

Comment:
I think, it should stay in the `NormalJoinTest`, because it depends on `joinFunction` and the form of the input tables. For example, see hashed join. Which result is only sorted, if the bigger input table was sorted.